src/apps/transfer/ICS20Transfer.sol

```diff
@@ -0,0 +1,141 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity >=0.8.25;
+
+import { IIBCApp } from "../../interfaces/IIBCApp.sol";
+import { IICS20Errors } from "./IICS20Errors.sol";
+import { ICS20Lib } from "./ICS20Lib.sol";
+import { IERC20 } from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import { SafeERC20 } from "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import { Ownable } from "@openzeppelin/contracts/access/Ownable.sol";
+import { ReentrancyGuard } from "@openzeppelin/contracts/utils/ReentrancyGuard.sol";
+import { IICS20Transfer } from "./IICS20Transfer.sol";
+
+using SafeERC20 for IERC20;
+
+/*
+ * Things not handled yet:
+ * - Prefixed denoms (source chain is not the source) and the burning of tokens related to that
+ * - Separate escrow balance tracking
+ * - Related to escrow ^: invariant checking (where to implement that)?
+ * - Receiving packets
+ */
+contract ICS20Transfer is IIBCApp, IICS20Transfer, IICS20Errors, Ownable, ReentrancyGuard {
+    string public constant ICS20_VERSION = "ics20-1";
+
+    /// @param owner_ The owner of the contract
+    constructor(address owner_) Ownable(owner_) { }
+
+    function onSendPacket(OnSendPacketCallback calldata msg_) external override onlyOwner nonReentrant {
+        if (keccak256(abi.encodePacked(msg_.packet.version)) != keccak256(abi.encodePacked(ICS20_VERSION))) {
+            revert ICS20UnexpectedVersion(msg_.packet.version);
+        }
+
+        UnwrappedFungibleTokenPacketData memory data = _unwrapPacketData(msg_.packet.data);
+
+        // TODO: Maybe have a "ValidateBasic" type of function that checks the packet data, could be done in unwrapping?
+
+        if (data.amount == 0) {
```

Comment:
This is a good question. There are two issues here I suppose.

- One is the actual uint256 which might cause problems (can be too big for go, I guess?)
- The other is the decimals variable itself, which is mostly required for frontends to make sense of a token. If the contract implements `IERC20Metadata` we can get this info, but not sure if we can do much with it. Even if the number of decimals makes sense, the number might still be too big.

One thing we could do is to check the size of the number and if it would not fit into go uint64, we would reject? Not sure what the alternative is, if the Cosmos SDK and ibc-go doesn't support big numbers. 

I saw someone mentioning this problem, but not sure where things break: https://x.com/brane_trix/status/1818055801463812494

Reply:
I think we can handle this situation if we know the decimals variable. Can discuss this later I suppose